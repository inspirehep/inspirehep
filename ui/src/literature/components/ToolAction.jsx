import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { ToolOutlined } from '@ant-design/icons';
import { Button } from 'antd';

import DropdownMenu from '../../common/components/DropdownMenu';
import IconText from '../../common/components/IconText';
import UserAction from '../../common/components/UserAction';
import { MAX_BULK_ASSIGN } from '../constants';
import ExportToCdsModal from './ExportToCdsModal';

function ToolAction({
  onAssignToConference,
  onExportToCds,
  disabledBulkAssign,
  selectionSize,
}) {
  const [isModalVisible, setIsModalVisible] = useState(false);
  const onClickExportToCds = useCallback(() => {
    setIsModalVisible(true);
  }, []);
  const onExportToCdsModalCancel = useCallback(() => {
    setIsModalVisible(false);
  }, []);

  const handleCdsExportOk = () => {
    setIsModalVisible(false);
    onExportToCds();
  };

  const menuItems = [
    {
      key: 'assign-conference',
      label: (
        <span
          title={
            disabledBulkAssign
              ? `Please select up to ${MAX_BULK_ASSIGN} papers that you want to assign to a conference.`
              : null
          }
          disabled={disabledBulkAssign}
          data-test-id="assign-conference"
          data-testid="assign-conference"
          onClick={() => onAssignToConference()}
        >
          Assign conference
        </span>
      ),
    },
    {
      key: 'export-to-CDS',
      label: (
        <span
          title={
            disabledBulkAssign
              ? `Please select up to ${MAX_BULK_ASSIGN} papers that you want to export to CDS.`
              : null
          }
          disabled={disabledBulkAssign}
          data-test-id="export-to-CDS"
          data-testid="export-to-CDS"
          onClick={onClickExportToCds}
        >
          Export to CDS
        </span>
      ),
    },
  ];

  return (
    <UserAction>
      <DropdownMenu
        title={
          <Button>
            <IconText text="tools" icon={<ToolOutlined />} />
          </Button>
        }
        items={menuItems}
      />
      <ExportToCdsModal
        onCancel={onExportToCdsModalCancel}
        onOk={handleCdsExportOk}
        visible={isModalVisible}
        selectionSize={selectionSize}
      />
    </UserAction>
  );
}

ToolAction.propTypes = {
  onAssignToConference: PropTypes.func.isRequired,
  onExportToCds: PropTypes.func.isRequired,
  disabledBulkAssign: PropTypes.bool,
  selectionSize: PropTypes.number,
};

export default ToolAction;
