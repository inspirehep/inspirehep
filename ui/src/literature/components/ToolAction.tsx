import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { ToolOutlined } from '@ant-design/icons';
import { Button, Menu } from 'antd';

import DropdownMenu from '../../common/components/DropdownMenu';
import IconText from '../../common/components/IconText';
import ListItemAction from '../../common/components/ListItemAction';
import { MAX_BULK_ASSIGN } from '../constants';
import ExportToCdsModal from './ExportToCdsModal';

function ToolAction({
  onAssignToConference,
  onExportToCds,
  disabledBulkAssign,
  selectionSize,
}) {
  const [isModalVisible, setIsModalVisible] = useState(false);
  const onClickExportToCds = useCallback(() => {
    setIsModalVisible(true);
  }, []);
  const onExportToCdsModalCancel = useCallback(() => {
    setIsModalVisible(false);
  }, []);

  const handleCdsExportOk = () => {
    setIsModalVisible(false);
    onExportToCds();
  };

  return (
    <ListItemAction>
      <DropdownMenu
        title={
          <Button>
            <IconText text="tools" icon={<ToolOutlined />} />
          </Button>
        }
      >
        <Menu.Item
          title={
            disabledBulkAssign
              ? `Please select up to ${MAX_BULK_ASSIGN} papers that you want to assign to a conference.`
              : null
          }
          disabled={disabledBulkAssign}
          data-test-id="assign-conference"
          key="assign-conference"
          onClick={() => onAssignToConference()}
        >
          Assign conference
        </Menu.Item>
        <Menu.Item
          title={
            disabledBulkAssign
              ? `Please select up to ${MAX_BULK_ASSIGN} papers that you want to export to CDS.`
              : null
          }
          disabled={disabledBulkAssign}
          data-test-id="export-to-CDS"
          key="export-to-CDS"
          onClick={onClickExportToCds}
        >
          Export to CDS
        </Menu.Item>
        <ExportToCdsModal
          onCancel={onExportToCdsModalCancel}
          onOk={handleCdsExportOk}
          visible={isModalVisible}
          selectionSize={selectionSize}
        />
      </DropdownMenu>
    </ListItemAction>
  );
}

ToolAction.propTypes = {
  onAssignToConference: PropTypes.func.isRequired,
  onExportToCds: PropTypes.func.isRequired,
  disabledBulkAssign: PropTypes.bool,
  selectionSize: PropTypes.number,
};

export default ToolAction;
