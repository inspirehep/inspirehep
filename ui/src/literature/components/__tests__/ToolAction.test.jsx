import React from 'react';
import { render, waitFor, fireEvent, screen } from '@testing-library/react';
import { shallow } from 'enzyme';

import ToolAction from '../ToolAction';

jest.mock('react-router-dom', () => ({
  useParams: jest.fn().mockImplementation(() => ({
    id: 123,
  })),
}));

describe('ToolAction', () => {
  it('renders', () => {
    const wrapper = shallow(
      <ToolAction
        onAssignToConference={jest.fn()}
        onExportToCds={jest.fn()}
        selectionSize={3}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const wrapper = shallow(
      <ToolAction
        onAssignToConference={jest.fn()}
        disabledBulkAssign
        onExportToCds={jest.fn()}
        selectionSize={3}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onAssignToConference on assign-conference click ', async () => {
    const onAssignToConference = jest.fn();
    const { container } = render(
      <ToolAction
        onAssignToConference={onAssignToConference}
        onExportToCds={jest.fn()}
        selectionSize={3}
      />
    );

    const dropdown = container.getElementsByClassName(
      'ant-dropdown-trigger'
    )[0];

    await waitFor(() => fireEvent.mouseOver(dropdown));
    await waitFor(() => screen.getByTestId('assign-conference').click());

    await waitFor(() => expect(onAssignToConference).toHaveBeenCalled());
  });

  it('opens modal on export-to-CDS click ', async () => {
    const onAssignToConference = jest.fn();
    const onExportToCds = jest.fn();

    const { container } = render(
      <ToolAction
        onAssignToConference={onAssignToConference}
        onExportToCds={onExportToCds}
        selectionSize={3}
      />
    );

    const dropdown = container.getElementsByClassName(
      'ant-dropdown-trigger'
    )[0];

    await waitFor(() => fireEvent.mouseOver(dropdown));
    await waitFor(() => screen.getByTestId('export-to-CDS').click());

    await waitFor(
      () => expect(screen.getByTestId('export-modal')).toBeInTheDocument
    );
  });
});
