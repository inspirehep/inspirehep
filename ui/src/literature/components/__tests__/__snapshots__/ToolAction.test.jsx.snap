// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolAction renders 1`] = `
<ListItemAction>
  <DropdownMenu
    disabled={false}
    title={
      <ForwardRef(InternalButton)>
        <IconText
          icon={<ForwardRef(ToolOutlined) />}
          text="tools"
        />
      </ForwardRef(InternalButton)>
    }
  >
    <MenuItem
      data-test-id="assign-conference"
      key="assign-conference"
      onClick={[Function]}
      title={null}
    >
      Assign conference
    </MenuItem>
    <MenuItem
      data-test-id="export-to-CDS"
      key="export-to-CDS"
      onClick={[Function]}
      title={null}
    >
      Export to CDS
    </MenuItem>
    <ExportToCdsModal
      onCancel={[Function]}
      onOk={[Function]}
      selectionSize={3}
      visible={false}
    />
  </DropdownMenu>
</ListItemAction>
`;

exports[`ToolAction renders disabled 1`] = `
<ListItemAction>
  <DropdownMenu
    disabled={false}
    title={
      <ForwardRef(InternalButton)>
        <IconText
          icon={<ForwardRef(ToolOutlined) />}
          text="tools"
        />
      </ForwardRef(InternalButton)>
    }
  >
    <MenuItem
      data-test-id="assign-conference"
      disabled={true}
      key="assign-conference"
      onClick={[Function]}
      title="Please select up to 250 papers that you want to assign to a conference."
    >
      Assign conference
    </MenuItem>
    <MenuItem
      data-test-id="export-to-CDS"
      disabled={true}
      key="export-to-CDS"
      onClick={[Function]}
      title="Please select up to 250 papers that you want to export to CDS."
    >
      Export to CDS
    </MenuItem>
    <ExportToCdsModal
      onCancel={[Function]}
      onOk={[Function]}
      selectionSize={3}
      visible={false}
    />
  </DropdownMenu>
</ListItemAction>
`;
